src/ik_memetic.cpp

```diff
@@ -241,21 +241,118 @@ auto ik_memetic(std::vector<double> const& initial_guess,
         if (print_debug) fmt::print("Returning best solution\n");
-        return ik.best().genes;
+        return ik.best();
+    }
+
+    return std::nullopt;
+}
+
+auto ik_memetic(std::vector<double> const& initial_guess,
+                Robot const& robot,
+                CostFn const& cost_fn,
+                SolutionTestFn const& solution_fn,
+                MemeticIkParams const& params,
+                double const timeout,
+                bool const approx_solution,
+                bool const print_debug) -> std::optional<std::vector<double>> {
+    // Check whether the initial guess already meets the goal,
+    // before starting to solve.
+    if (solution_fn(initial_guess)) {
+        return initial_guess;
     }
 
+    std::atomic<bool> terminate{false};
+    if (params.num_threads <= 1) {
+        // Single-threaded implementation
+        auto maybe_solution = ik_memetic_impl(initial_guess,
+                                              robot,
+                                              cost_fn,
+                                              solution_fn,
+                                              params,
+                                              terminate,
+                                              timeout,
+                                              approx_solution,
+                                              print_debug);
+        if (maybe_solution.has_value()) {
+            return maybe_solution.value().genes;
+        }
+    } else {
+        // Multi-threaded implementation
+        rsl::Queue<std::optional<Individual>> solution_queue;
+        std::vector<std::thread> ik_threads;
+        ik_threads.reserve(params.num_threads);
+
+        auto ik_thread_fn = [=, &terminate, &solution_queue]() {
+            auto soln = ik_memetic_impl(initial_guess,
+                                        robot,
+                                        cost_fn,
+                                        solution_fn,
+                                        params,
+                                        terminate,
+                                        timeout,
+                                        approx_solution,
+                                        print_debug);
+            solution_queue.push(soln);
+        };
+
+        for (size_t i = 0; i < params.num_threads; ++i) {
+            ik_threads.push_back(std::thread(ik_thread_fn));
+        }
+
+        // If enabled, stop on first solution and return.
+        size_t num_done = 0;
+        std::vector<double> best_solution;
+        auto min_cost = std::numeric_limits<double>::max();
+        if (params.stop_on_first_soln) {
+            while (solution_queue.empty() && !terminate && num_done < params.num_threads) {
+                std::this_thread::sleep_for(std::chrono::microseconds(100));
+            }
+            auto const maybe_solution = solution_queue.pop().value();
+            if (maybe_solution.has_value()) {
```

Comment:
Once we get here (if you've removed the above line) this just becomes:
```suggestion
            if (maybe_solution.value().has_value()) {
```